### PR TITLE
StateTimeline: Share cursor with rest of the panels

### LINKED ISSUE #35908: Regression: Shared crosshair works only for panels of the same type
**What happened**:
Shared crosshair support partially broke somewhere between v7.0.6 (working) and >= v7.5.6 (broken). Cannot tell exactly which version broke it, but now it's only synced between **panels of the same type**, while other types don't show the line anymore.

**What you expected to happen**:
In v7.0.6, shared crosshair worked across panels of all types, e.g. 'graph' & 'discrete' next to each-other.

**How to reproduce it (as minimally and precisely as possible)**:
Enable shaired crosshair and have at least two panels of different types, e.g. 'graph' & 'discrete'.

**Anything else we need to know?**:
Broken in v8.0.2, too, same applies to the new/beta panel types.

**Environment**:
- Grafana version: v7.5.6, v8.0.2
- Data source type & version: Doesn't matter
- OS Grafana is installed on: Linux/Docker
- User OS & Browser: Windows 10 Firefox 89
- Grafana plugins: 'discrete' panel

### PR DESCRIPTION
**Pull request by zoltanbedi**
**What this PR does / why we need it**:

This PR adds shared cursor support for the state timeline panel.

https://user-images.githubusercontent.com/13729989/139251511-692d7f66-8f37-4564-bfce-41bb767ca904.mp4

## TODO

- [x] Performance issue / jumpy tooltip when hovered over state timeline
- [x] Show tooltip of the first serie always?

**Which issue(s) this PR fixes**:

Fixes #35908 https://github.com/grafana/support-escalations/issues/934

<details>
<summary>Dashboard JSON - with gdev graphite ds</summary>

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "gnetId": null,
  "graphTooltip": 2,
  "id": 178,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "backgroundColor": "rgba(128,128,128,0.1)",
      "colorMaps": [
        {
          "$$hashKey": "object:4633",
          "color": "#CCC",
          "text": "N/A"
        }
      ],
      "crosshairColor": "#8F070C",
      "datasource": "gdev-graphite",
      "display": "timeline",
      "expandFromQueryS": 0,
      "extendLastValue": true,
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "highlightOnMouseover": true,
      "id": 8,
      "legendSortBy": "-ms",
      "lineColor": "rgba(0,0,0,0.1)",
      "metricNameColor": "#000000",
      "rangeMaps": [
        {
          "$$hashKey": "object:4643",
          "from": "null",
          "text": "N/A",
          "to": "null"
        }
      ],
      "rowHeight": 50,
      "showDistinctCount": false,
      "showLegend": true,
      "showLegendCounts": false,
      "showLegendNames": false,
      "showLegendPercent": true,
      "showLegendTime": false,
      "showLegendValues": true,
      "showTimeAxis": true,
      "showTransitionCount": false,
      "targets": [
        {
          "datasource": "gdev-graphite",
          "refId": "A",
          "target": "aliasByNode(stats_counts.*.*, 2)"
        }
      ],
      "textSize": 24,
      "textSizeTime": 12,
      "timeOptions": [
        {
          "name": "Years",
          "value": "years"
        },
        {
          "name": "Months",
          "value": "months"
        },
        {
          "name": "Weeks",
          "value": "weeks"
        },
        {
          "name": "Days",
          "value": "days"
        },
        {
          "name": "Hours",
          "value": "hours"
        },
        {
          "name": "Minutes",
          "value": "minutes"
        },
        {
          "name": "Seconds",
          "value": "seconds"
        },
        {
          "name": "Milliseconds",
          "value": "milliseconds"
        }
      ],
      "timePrecision": {
        "name": "Minutes",
        "value": "minutes"
      },
      "timeTextColor": "#d8d9da",
      "title": "Discrete",
      "type": "natel-discrete-panel",
      "units": "short",
      "use12HourClock": false,
      "useTimePrecision": false,
      "valueMaps": [
        {
          "$$hashKey": "object:4641",
          "op": "=",
          "text": "N/A",
          "value": "null"
        }
      ],
      "valueTextColor": "#000000",
      "writeAllValues": false,
      "writeLastValue": false,
      "writeMetricNames": true
    },
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "gdev-graphite",
      "fieldConfig": {
        "defaults": {},
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "A-series"
            },
            "properties": [
              {
                "id": "unit",
                "value": "none"
              }
            ]
          },
          {
            "matcher": {
              "id": "byName",
              "options": "B-series"
            },
            "properties": [
              {
                "id": "unit",
                "value": "none"
              }
            ]
          },
          {
            "matcher": {
              "id": "byName",
              "options": "C-series"
            },
            "properties": [
              {
                "id": "unit",
                "value": "none"
              }
            ]
          }
        ]
      },
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 2,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "alertThreshold": true
      },
      "percentage": false,
      "pluginVersion": "8.3.0-pre",
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "datasource": "gdev-graphite",
          "refId": "A",
          "target": "aliasByNode(stats_counts.*.*, 2)"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "Old Graph",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "$$hashKey": "object:81",
          "format": "none",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "$$hashKey": "object:82",
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    },
    {
      "datasource": "gdev-graphite",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 70,
            "lineWidth": 0
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "light-blue",
                "value": null
              },
              {
                "color": "semi-dark-green",
                "value": 5
              },
              {
                "color": "semi-dark-yellow",
                "value": 10
              },
              {
                "color": "semi-dark-red",
                "value": 20
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 8
      },
      "id": 6,
      "options": {
        "alignValue": "left",
        "legend": {
          "displayMode": "list",
          "placement": "bottom"
        },
        "mergeValues": false,
        "rowHeight": 0.9,
        "showValue": "auto",
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "datasource": "gdev-graphite",
          "refId": "A",
          "target": "aliasByNode(stats_counts.*.*, 2)"
        }
      ],
      "title": "State timeline",
      "type": "state-timeline"
    },
    {
      "datasource": "gdev-graphite",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 12,
        "y": 8
      },
      "id": 4,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi"
        }
      },
      "targets": [
        {
          "datasource": "gdev-graphite",
          "refId": "A",
          "target": "aliasByNode(stats_counts.*.*, 2)"
        }
      ],
      "title": "Time series",
      "type": "timeseries"
    }
  ],
  "refresh": false,
  "schemaVersion": 32,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-1h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "shared",
  "uid": "BtChWlink",
  "version": 15,
  "weekStart": ""
}
```
</details>

### CODE REVIEW

**Comment by zoltanbedi** [comment]
Looks like the "performance issue" is in my google chrome for some reason (probably a plugin). In firefox it looks good.

https://user-images.githubusercontent.com/13729989/140086435-096b0a01-2cb8-4317-a140-00dca227979d.mp4

File: packages/grafana-ui/src/components/uPlot/utils.ts [hunk 1]
```diff
@@ -181,6 +181,11 @@ export function findMidPointYPosition(u: uPlot, idx: number) {
     y = u.valToPos((min || max)!, u.series[(sMaxIdx || sMinIdx)!].scale!);
   }
 
+  // if y is out of canvas bounds, snap it to the bottom
+  if (y !== undefined && y < 0) {
+    y = u.bbox.height;
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/utils.ts:186]: `bbox` is in canvas pixels, so i think this needs a `/ devicePixelRatio`

(`u.valToPos()` returns CSS pixels, unless third arg is `true` for canvas pixels)

Comment (zoltanbedi) [reply packages/grafana-ui/src/components/uPlot/utils.ts:186]: You are right this needed the devicePixelRatio to work properly.

**Review by dprokop — APPROVED** [review]
Think this is a great improvement in current shape!

Can already think about cursor sync being slightly more clever in the future when syncing timeline with other panels:
- When graph tooltip in single mode: sync is series-aware so that when you hover over A-series then cursor is snapped to A-series in other panels 
- When graph tooltip in multi-mode: only vertical crosshair is shown on other panels and the cursor syncs as if it followed only the time axis rather than position within the visualization.

But these are improvements we can think about in future PRs.

**Comment by mrkeuz** [comment]
Hi really wait this feature (State Timeline charts). Thanks for PR and your work. 🚀🚀🚀

But seems working now well. Found two issues on v8.3.0-beta1 (fe082e2210) (docker version)

1. From State Timeline charts crosshair work well but not vice-versa. But cursor appear sometime but hang in random places (see screencast)
2. Empty (???) charts with alerts (???) not show crosshair sometimes, (bottom right on screencast)

Can provide additional info or testing something.

StateTimeline_missing_cursor_sync

**Comment by zoltanbedi** [comment]
Hi @mrkeuz,
I'm happy that you were able to try it out and let us know about possible issues. Can you please open a new GitHub issue with your findings please se we can track it properly?
Thank you

**Comment by mrkeuz** [comment]
@zoltanbedi, done, issue added #42102

**Pull request merged**